src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -5404,7 +5404,16 @@ private static (BoundExpression expression, Conversion conversion) RemoveConvers
                 {
                     break;
                 }
+
                 var conversion = (BoundConversion)expr;
+
+                if (avoidTypelessExpressions &&
+                    conversion.ConversionKind == ConversionKind.ImplicitReference &&
+                    conversion.Operand is BoundLiteral { Type: null })
```

Comment:
It feels like the current behavior of `null` is expected: we pass an untyped `BoundLiteral` to `MethodTypeInferrer`. But the behavior of `default` is unexpected: we pass a typed `BoundDefaultExpression` to `MethodTypeInferrer` where the type was only determined after running `MethodTypeInferrer` in initial binding.

Reply:
Just to clarify, you're saying we should keep the current behavior for `null` (`Infer("", null)` would infer `string!` and warn, rather than `string?`)?

Reply:
Not necessarily. I'm suggesting the arguments passed to `MethodTypeInferrer` in `NullableWalker` should match the typed/untyped-ness of the arguments passed to `MethodTypeInferrer` in initial binding. So `null` and `default` would both be passed as untyped arguments. Perhaps `NullableWalker` could fix up the (top-level?) nullability of the inferred type arguments after `MethodTypeInferrer` returns.

---
In reply to: [416076359](https://github.com/dotnet/roslyn/pull/43571#discussion_r416076359) [](ancestors = 416076359)